async-http-client/src/main/scala/org/htt4s/client/asynchttpclient/AsyncHttpClient.scala

```diff
@@ -0,0 +1,102 @@
+package org.htt4s.client.asynchttpclient
+
+import com.ning.http.client.AsyncHandler.STATE
+import com.ning.http.client.generators.InputStreamBodyGenerator
+import com.ning.http.client.{Request => AsyncRequest, Response => AsyncResponse, _}
+
+import org.http4s._
+import org.http4s.client._
+import org.http4s.util.task
+import scodec.bits.ByteVector
+
+import scala.collection.JavaConverters._
+
+import scalaz.stream.io._
+import scalaz.stream.async
+import scalaz.concurrent.Task
+
+import scala.concurrent.Promise
+import scala.concurrent.ExecutionContext.Implicits.global
+
+object AsyncHttpClient {
+  val EOF = ByteVector(-1)
+
+  val defaultConfig = new AsyncHttpClientConfig.Builder()
+    .setMaxConnectionsPerHost(200)
+    .setMaxConnections(400)
+    .setRequestTimeout(30000)
+    .build()
+
+  def apply(config: AsyncHttpClientConfig = defaultConfig): Client = {
+    val client = new AsyncHttpClient(config)
+    Client(Service.lift { req =>
+      val p = Promise[DisposableResponse]
+      client.executeRequest(toAsyncRequest(req), asyncHandler(p))
+      task.futureToTask(p.future)
+    }, Task(client.close()))
+  }
+
+  private def asyncHandler(promise: Promise[DisposableResponse]): AsyncHandler[Unit] =
+    new AsyncHandler[Unit] {
+      var state: AsyncHandler.STATE = STATE.CONTINUE
+      val queue = async.unboundedQueue[ByteVector]
```

Comment:
There is a bounded queue.  The enqueue task does not complete until there is space in the queue, which will means the thread that calls `onBodyPartReceived` will be blocked with a slow consumer.

Sizing the bounded queue is tricky too, because the bound is in chunk count, not byte count.  If the chunks are REALLY_BIG and the queue is size N, we might buffer N * REALLY_BIG bytes.